src/ILCompiler.CppCodeGen/src/CppCodeGen/CppWriter.cs

```diff
@@ -1386,6 +2242,92 @@ private void OutputUnboxingStubNode(CppUnboxingStubNode unboxingStubNode)
             Out.Write(sb.ToString());
         }
 
+        private void OutputStaticsCode(Dictionary<TypeDesc, CppGenerationBuffer> statics,
+            bool isGCStatic = false, bool isThreadStatic = false)
+        {
+            CppGenerationBuffer sb = new CppGenerationBuffer();
+
+            foreach (var entry in statics)
+            {
+                TypeDesc t = entry.Key;
+
+                sb.Append("struct ");
+                sb.Append(GetCppStaticsTypeName(t, isGCStatic, isThreadStatic));
+                sb.Append(" {");
+                sb.Append(entry.Value.ToString());
+                sb.Exdent();
+                sb.AppendLine();
+                sb.Append("};");
+                sb.AppendLine();
+
+                if (t.IsCanonicalSubtype(CanonicalFormKind.Any))
+                    continue;
+
+                sb.Append(GetCppStaticsTypeName(t, isGCStatic, isThreadStatic));
+                sb.Append(" ");
+                sb.Append(GetCppStaticsName(t, isGCStatic, isThreadStatic));
+
+                if (!isGCStatic && t.HasStaticConstructor)
+                {
+                    MethodDesc cctor = t.GetStaticConstructor();
+                    MethodDesc canonCctor = cctor.GetCanonMethodTarget(CanonicalFormKind.Specific);
+
+                    if (_typesWithCctor.Contains(canonCctor.OwningType))
+                    {
+                        sb.Append(" = {");
+                        sb.Indent();
+                        sb.AppendLine();
+
+                        if (canonCctor.RequiresInstArg())
+                        {
+                            sb.Append(".__cctor = (void (*)(void *)) *(void **)");
+                            sb.Append(GetCppFatFunctionPointerNameForMethod(cctor));
+                            sb.Append("(),");
+                            sb.Append(".__ctx = **(void***) (((intptr_t)");
+                            sb.Append(GetCppFatFunctionPointerNameForMethod(cctor));
+                            sb.Append("()) + sizeof(void*))");
+                        }
+                        else
+                        {
+                            sb.Append(".__cctor = &");
+                            sb.Append(GetCppMethodDeclarationName(canonCctor.OwningType, GetCppMethodName(canonCctor)));
+                        }
+
+                        sb.Exdent();
+                        sb.AppendLine();
+                        sb.Append("}");
+                    }
+                }
+
+                sb.Append(";");
+                sb.AppendLine();
+            }
+
+            Out.Write(sb.ToString());
+        }
+
+        public TypeDesc ConvertToCanonFormIfNecessary(TypeDesc type, CanonicalFormKind policy)
```

Comment:
I think this can just be:

```csharp
public static TypeDesc CanonicallyNormalized(this TypeDesc type)
{
    if (type.IsCanonicalSubtype(CanonicalFormKind.Any))
        return type.ConvertToCanonForm(CanonicalFormKind.Specific);
    return type;
}
```

The pointers and byrefs shouldn't require special casing, and we don't need to specify canonical form as a parameter (if the type happens to have `CanonicalFormKind.Universal` types in it, the `ConvertToCanonForm` will to the necessary upgrade).

This method could also be static/extension method, and we could place it in TypeExtensions.cs. I know at least one more place that could use it.

Reply:
I used special casing for pointers and byrefs to do conversions like this: `Dictionary<String, Canon>&` -> `Dictionary<Canon, Canon>&`. `ConvertToCanonForm` converts such types to `Canon&` that isn't convenient for cpp codegen.

Reply:
I see. CppCodegen requires different canonical forms than what I'm used to. This looks good.